### PR TITLE
AirfRANS: 4L/256d gc=1.5 + T_max=5 — architecture-enabled gradient clipping

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The 4L/256d architecture's weight decay stabilization (WD=1e-2) may enable gc=1.5 where it previously caused volatility at 3L/192d. The triple compound (gc=1.5 + T_max=5 + WD=1e-2) failed at smaller architecture but at 4L/256d the richer representational capacity combined with WD regularization should absorb the larger gradient steps without divergence. gc=1.5 was the previous best on AirfRANS at 0.00935 — applying it to the 4L/256d architecture that hit 0.007264 is the highest-priority next step.

## Instructions

Start from the golden 4L/256d command and increase grad-clip from 1.0 to 1.5. Everything else stays identical:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-gc
```

**Key change:** --grad-clip 1.5 (was --grad-clip 1.0)

Monitor grad norm behavior carefully in W&B. At 3L/192d gc=1.5 caused volatility — check if 4L/256d absorbs it. Report the grad norm trajectory alongside val/surface_mse.

## Baseline

- **Primary metric:** val_primary/surface_mse
- **Current best:** 0.007264 (val) at epoch 40
- **Best PR:** #2727 (shoya — 4L/256d + gc=1.0 + WD=1e-2 + T_max=5, 50 epochs, hit epoch cap at 61 min of 180-min budget)
- **External target:** 0.0043 (~1.7x gap remaining)
- **W&B run:** ruurxdqs
- **Reproduce baseline:** cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

**Verdict: NEGATIVE — gc=1.5 does not beat gc=1.0 at 4L/256d**

| Metric | Baseline (gc=1.0) | This run (gc=1.5) | Delta |
|--------|-------------------|-------------------|-------|
| val_primary/surface_mse (best) | **0.007264** (ep40) | 0.01438 (ep35) | +98% (worse) |
| test_primary/surface_mse | — | 0.02616 (ep36, not at best val) | — |
| full_val/volume_mse (best) | — | 0.0652 (ep36) | — |
| Epochs | 50 (61 min) | 36 (30 min timeout) | — |

### Cosine trough envelope (T_max=5, troughs at multiples of 5)

| Epoch | val_primary/surface_mse |
|-------|------------------------|
| 5 | 0.2332 |
| 10 | 0.1504 |
| 15 | 0.0576 |
| 20 | 0.0461 |
| 25 | 0.0953 (anomalous spike) |
| 30 | 0.0248 |
| **35** | **0.0144** (best) |

Trough envelope still descending but would need ~0.007 at ep40 to match baseline — unlikely given the descent rate.

### Grad norm trajectory (gc=1.5 stability check)

| Epoch | grad_norm_mean | grad_clip_events (of 360) |
|-------|---------------|--------------------------|
| 1 | 10.5 | 356 (99%) |
| 7 | 18.5 | 352 (98%) |
| 14 | **26.6** (peak) | 342 (95%) |
| 20 | 18.3 | 321 (89%) |
| 30 | 17.4 | 319 (89%) |
| 35 | 13.5 | 295 (82%) |
| 36 | 12.2 | 276 (77%) |

Grad norms peaked at epoch 14 (26.6) then steadily decreased to ~12. No divergence — the 4L/256d architecture **does** absorb gc=1.5 (hypothesis partially validated). Clip events decreased from 99% to 77%, confirming gradients are becoming more controlled as training progresses.

### What happened

The 4L/256d architecture successfully absorbs gc=1.5 without divergence — no volatility or gradient explosion. However, the looser clip threshold allows larger gradient steps that create more oscillation between cosine troughs, resulting in **slower convergence** compared to gc=1.0. At the same epoch count, gc=1.5 is roughly 2x worse. Even with more epochs, projecting the descent rate suggests gc=1.5 would reach ~0.008-0.010 at epoch 40, likely NOT matching gc=1.0's 0.007264.

The key insight: at 3L/192d, gc=1.5 was the **best** clip value (0.00935 vs gc=1.0's 0.01419). But at 4L/256d, gc=1.0 is already optimal — the larger architecture doesn't benefit from the looser clipping. The WD=1e-2 regularization provides sufficient gradient control that tighter clipping (1.0) is preferred.

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --agent fern --wandb-name "fern/airfrans-4L256d-gc15-tmax5" \
  --wandb-group airfrans-4L256d-gc
```

- **W&B run:** jugzoyhb
- **Peak memory:** Not logged (GPU memory freed after run)

### Suggested follow-ups

1. **gc=1.2 at 4L/256d** — interpolate between gc=1.0 (baseline) and gc=1.5 to find the sweet spot
2. **gc=1.5 with longer training** — run with 180-min budget to see if the trough envelope catches up at epoch 50+
3. **gc=0.8 at 4L/256d** — explore whether tighter clipping improves 4L/256d further (given gc=1.0 > gc=1.5, maybe gc<1.0 is even better)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2766 because ADVISOR: Closing — gc=1.5 at 4L/256d achieved 0.014384, nearly 2x worse than baseline. This used WD=1e-2 which we've confirmed fights gc=1.5. We have cleaner gc=1.5 tests running without WD (haku #2791, edward #2785, nezuko #2784). Thank you for the data point..
